File: src/tuk-cps/main/java/ExprParser/ExprTreeBinOp.java
package ExprParser;

/**
 * A binary operation.
 */
class ExprTreeBinOp extends ExprTree {
    ExprTree l;
    ExprTree r;
    int op;

    ExprTreeBinOp(ExprTree l, int op, ExprTree r)  {
        super(l.result);
        assert l != null && r != null: "Operands must not be null";
        this.l = l.copy(); this.op = op; this.r = r.copy();
    }

    ExprTree copy() {
        ExprTreeBinOp cp = new ExprTreeBinOp(this.l, this.op, this.r);
        cp.setId(this.getId());
        cp.setResult(this.getResult());
        return cp;
    }


    // Computes the complete tree recursively.
    void evalExpr() throws ExprError {
        l.evalExpr();
        r.evalExpr();
        evalThis();
        return;
    }


    // Just compute operation with two operands: l and r.
    void evalThis() throws ExprError {
        switch(op) {
            case '+': result = l.getAADD().add(r.getAADD()); break;
            case '-': result = l.getAADD().sub(r.getAADD()); break;
            case '*': result = l.getAADD().mult(r.getAADD()); break;
            case '/': result = l.getAADD().div(r.getAADD()); break;
            case '>': result = l.getAADD().Gt(r.getAADD()); break;
            case '<': result = l.getAADD().Lt(r.getAADD()); break;
            case ExprScanner.GE: result = l.getAADD().Ge(r.getAADD()); break;
            case ExprScanner.LE: result = l.getAADD().Le(r.getAADD()); break;
            case '&': result = l.getBDD().and(r.getBDD()); break;
            case '|': result = l.getBDD().or(r.getBDD()); break;
            case '=': throw new ExprError("Equality of two AADD is not supported.");
            default:  throw new ExprError("Operation "+ (char) op + " resp. " + op + " not supported on AADD.");
        }
    }


    public String toString() {
        return "(" + l.toString()+(char) op +r.toString() + ")";
    }
}

File: src/main/kotlin/com/github/tuk-cps/exprParser/ExprTreeBinOp.kt
package exprParser

/**
 * @class ExprTreeBinOp
 * A binary operation.
 *
 * @author Christoph Grimm, Jack D. Martin
 */
internal class ExprTreeBinOp(l: ExprTree, op: Int, r: ExprTree) : ExprTree(l.value, "internal")  {

    var l: ExprTree = l.copy()
    var r: ExprTree = r.copy()
    var op = op

    init {
        l.parent = this
        r.parent = this
    }

    override fun copy(): ExprTreeBinOp {
        val cp = ExprTreeBinOp(l, op, r)
        cp.parent = parent
        cp.id = id
        cp.value = value
        return cp
    }

    /** Computes operands from leaves upwards */
    @Throws(ExprError::class)
    override fun evalUpRec() {
        l.evalUpRec()
        r.evalUpRec()
        evalUp()
    }

    /** Computes one level upwards, from childs to parent */
    @Throws(ExprError::class)
    override fun evalUp() {
        value = when (op) {
            '+'.toInt() -> l.aadd + r.aadd
            '-'.toInt() -> l.aadd - r.aadd
            '*'.toInt() -> l.aadd * r.aadd
            '/'.toInt() -> l.aadd / r.aadd
            '>'.toInt() -> l.aadd gt r.aadd
            '<'.toInt() -> l.aadd lt r.aadd
            ExprScanner.GE -> l.aadd ge r.aadd
            ExprScanner.LE -> l.aadd le r.aadd
            '&'.toInt() -> l.bdd and r.bdd
            '|'.toInt() -> l.bdd or r.bdd
            '='.toInt() -> throw ExprError("Equality of two AADD is not supported.")
            else -> throw ExprError("Operation " + op.toChar() + " resp. " + op + " not supported on AADD.")
        }
        println("computed up: $this")
    }

    /** Computes one level downwards, from parent to childs */
    override fun evalDown() {
        val prevL = l.aadd
        val prevR = r.aadd
        when (op) {
            '+'.toInt() -> { l.value = aadd - prevR; r.value = aadd - prevL }
            '-'.toInt() -> { l.value = aadd + prevR; r.value = aadd - prevL }
            '*'.toInt() -> { l.value = aadd / prevR; r.value = aadd / prevL }
            '/'.toInt() -> { l.value = aadd * prevR; r.value = prevL / aadd }
            else -> { }
        }
        l.aadd.intersect(prevL)
        r.aadd.intersect(prevR)
        println("computed down: $this")
    }

    override fun evalDownRec() {
        evalDown()
        l.evalDownRec()
        r.evalDownRec()
    }


    override fun toString() = "(id: $id $value ) = "+ l + op.toChar() + r
}